Frameworks/Foundation/NSIndexPath.mm

```diff
@@ -90,114 +112,100 @@ - (BOOL)isEqual:(id)other {
 /**
  @Status Interoperable
 */
-- (void)dealloc {
-    if (_indexes) {
-        IwFree(_indexes);
-    }
-    [super dealloc];
-}
-
-/**
- @Status Caveat
- @Notes position must be valid
-*/
-- (unsigned)indexAtPosition:(unsigned)position {
-    if (position >= _length) {
-        assert(0);
-        //[NSException raise:NSInvalidArgumentException format:@"Unable to remove index from zero length path."];
-        return 0;
-    }
-
-    return _indexes[position];
+- (NSUInteger)indexAtPosition:(NSUInteger)position {
+    return position < _indexes.size() ? _indexes[position] : NSNotFound;
 }
 
 /**
  @Status Interoperable
 */
 - (void)getIndexes:(NSUInteger*)indexes {
-    for (unsigned int i = 0; i < _length; i++) {
-        indexes[i] = _indexes[i];
-    }
+    std::copy(_indexes.cbegin(), _indexes.cend(), indexes);
 }
 
 /**
  @Status Interoperable
 */
 - (id)copyWithZone:(NSZone*)zone {
-    return [[[self class] alloc] initWithIndexes:_indexes length:_length];
+    return [[[self class] alloc] initWithIndexes:_indexes.data() length:_indexes.size()];
 }
 
 /**
  @Status Interoperable
 */
 - (NSIndexPath*)indexPathByAddingIndex:(NSUInteger)newIndex {
-    id ret = [[self class] alloc];
-
-    NSUInteger* indexCopy = (NSUInteger*)IwMalloc((_length + 1) * sizeof(DWORD));
-    memcpy(indexCopy, _indexes, _length * sizeof(DWORD));
-    indexCopy[_length] = newIndex;
-
-    [ret initWithIndexes:indexCopy length:_length + 1];
+    NSIndexPath* ret = [NSIndexPath indexPathWithIndexes:nullptr length:0];
+    if (ret) {
+        ret->_indexes.reserve(_indexes.size() + 1);
+        ret->_indexes.insert(ret->_indexes.begin(), _indexes.cbegin(), _indexes.cend());
+        ret->_indexes.emplace_back(newIndex);
+    }
 
-    IwFree(indexCopy);
+    return ret;
+}
 
-    return [ret autorelease];
+/**
+ @Status Interoperable
+*/
+- (NSIndexPath*)indexPathByRemovingLastIndex {
+    return [NSIndexPath indexPathWithIndexes:_indexes.data() length:(_indexes.empty() ? 0 : _indexes.size() - 1)];
 }
 
 /**
  @Status Interoperable
 */
 - (NSUInteger)length {
-    return _length;
+    return _indexes.size();
 }
 
 /**
  @Status Interoperable
 */
 - (NSUInteger)hash {
-    unsigned ret = 0;
-
-    for (unsigned int i = 0; i < _length; i++) {
-        ret += _indexes[i];
+    if ([self length] == 0L) {
+        return 0;
     }
 
-    return ret;
+    auto dataToHash = woc::MakeAutoCF<CFDataRef>(
+        CFDataCreateWithBytesNoCopy(nullptr, (const UInt8*)_indexes.data(), _indexes.size() * sizeof(NSUInteger), kCFAllocatorNull));
+    return CFHash(dataToHash);
 }
 
 /**
  @Status Interoperable
 */
-- (int)compare:(id)otherObj {
-    int len1 = [self length];
-    int len2 = [otherObj length];
+- (NSComparisonResult)compare:(id)otherObj {
```

Comment:
nit: if they are the same object, we can short circuit earlier vs doing more method calls and validating.